### PR TITLE
Dockerfile: do not use root processes (fix #72)

### LINKED ISSUE #72: Running UID/GID for the docker
I run the docker version of this and over ride the UID/GID with docker run command. However this is causing permission problem when the docker try to download the latest youtube-dl package.

As it is a common practice to not run docker as root user, the fixed could be changing the write permission of "modules/youtube-dl/bin/" folder.

internal/streams/legacy.js:59
      throw er; // Unhandled stream error in pipe.
      ^

[Error: EACCES: permission denied, open 'node_modules/youtube-dl/bin/youtube-dl'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: 'node_modules/youtube-dl/bin/youtube-dl'
}

### PR DESCRIPTION
**Pull request by Wonderfall**
Hi! First and foremost, I have to thank you for this great app.

Unfortunately, I was not at ease using your official Docker image which is running everything as `root` by default. I mean, it's possible to get rid of that issue with user namespaces, but that's a bit tricky (especially with volume points and all that stuff) for someone who just wants to deploy your app as soon as possible.

Therefore I've made some changes to your image and it runs flawlessly. I usually use `tini` in conjunction with `su-exec` to run everything under a non-root user (uid/gid can be set with `UID` and `GID` environment variables). To avoid permission issues, the startup script runs a recursive `chown` at startup (which can be slower on `overlay2` due to performance issues in this very specific case, otherwise it's not noticeable). It fixes #72.

It's a drop-in replacement and requires basically no changes to previous installations, but there has to be a note that permissions to the volumes will be changed accordingly, and that by default permissions will be changed for 991/991 as UID/GID (arbitrary choice).

I think it's something that needs to be addressed ASAP as running as root even in containers is considered a security bad practice, whether you like my solution or not. An alternative would be to specify a `USER` directive in the `Dockerfile`, which works too, but it's not as flexible and will cause frustration among Docker beginners.

By the way, if you merge this, I also upgraded the Dockerfile to Alpine 3.12 which was released some days ago. `python` is not a valid package anymore, so I had to specify `python2`.

Again, thanks for your work!

### CODE REVIEW

File: backend/Dockerfile [hunk 1]
```diff
@@ -15,4 +19,6 @@ COPY ./ /app/
 
 EXPOSE 17442
 
-CMD [ "node", "app.js" ]
+RUN chmod +x /usr/local/bin/run.sh
+
+CMD ["run.sh"]
```
Comment (SuperSandro2000) [inline backend/Dockerfile:24]: This should be an entrypoint.

File: backend/docker_run.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+chown -R $UID:$GID /app
+exec su-exec $UID:$GID /sbin/tini -- node app.js
```
Comment (SuperSandro2000) [inline backend/docker_run.sh:3]: This entrypoint should forward args if the cmd starts with - and call a binary otherwise like the official DockerHub Images do.

You can find an example of that [here](https://github.com/SuperSandro2000/docker-images/blob/master/reddiscord/files/entrypoint.sh).

File: backend/Dockerfile [hunk 3]
```diff
@@ -1,12 +1,16 @@
...
+ENV UID=991 GID=991
+
+RUN \
+apk add --no-cache npm python2 ffmpeg tini su-exec && \
 apk add --no-cache --repository http://dl-cdn.alpinelinux.org/alpine/edge/testing/ \
     atomicparsley
 
 WORKDIR /app
 
+COPY docker_run.sh /usr/local/bin/run.sh
```
Comment (SuperSandro2000) [inline backend/Dockerfile:12]: I think it would be less confusing if the file would be named locally run.sh. It is already in a docker directory so there is no confussion.

File: backend/Dockerfile [hunk 4]
```diff
@@ -1,12 +1,16 @@
-FROM alpine:3.11
+FROM alpine:3.12
 
-RUN \ 
-apk add --no-cache npm python ffmpeg && \
+ENV UID=991 GID=991
+
+RUN \
+apk add --no-cache npm python2 ffmpeg tini su-exec && \
```
Comment (SuperSandro2000) [inline backend/Dockerfile:6]: We do not need to install tini here. You can pass ``--init`` to docker run or similar to achieve the same result.

File: backend/Dockerfile [hunk 5]
```diff
@@ -1,12 +1,16 @@
-FROM alpine:3.11
+FROM alpine:3.12
```
Comment (SuperSandro2000) [inline backend/Dockerfile:1]: This update is totally unrelated to this pull request and should have been at least a different commit if not PR.

**Comment by Wonderfall** [comment]
I'll close this PR following the reviews. If you'd like to use this solution : 
- I agree `ENTRYPOINT`, while not strictly needed, could be used like this : 

```
ENTRYPOINT [ "run.sh" ]
CMD [ "node", "app.js" ]
```

```
exec su-exec $UID:$GID /sbin/tini "$@"
````

- I don't think advising `--init` §which I was aware of) is as user friendly. But I can see the redundancy somehow.

- Yeah sorry about upgrading Alpine in the same commit, I was just in a rush.

Either way, this works well for me and I hope it can inspire you to fix this. :)

**Comment by Tzahi12345** [comment]
Thank you @Wonderfall! I agree with @SuperSandro2000 that the PR should be limited to solving the root user issue, but it's great to see some help on this front (I'm not up to speed on how to make the image non-root friendly). I know it's a bad practice to have everything run as root, and it's why I'm happy to see it given some attention.

If you're willing to address some of the minor issues (I think tini support is native to Docker) and the arbitrary 991 UID (I'd prefer it be user customizable, perhaps with an ENV var as you mentioned), then I'd merge immediately :)

I do have a question since you've obviously given this more thought than me: would it be OK if by default it ran as the root user, but with the easy option of being able to set your own UID/GUID? Or is this also considered a bad practice and non-root should always be used?

I ask this because I don't want to override the permissions of folders for users and perhaps prevent access. I'm a bit weak on the ACL and UID permissions stuff so it's a bit of a knowledge gap on my end.

**Comment by Wonderfall** [comment]
Hi, thanks for your answer! I was just trying to solve the issue the way I did with my own Docker images as it suits my needs, but I understand very well if it's not convenient for others. :)

> If you're willing to address some of the minor issues (I think tini support is native to Docker) 

It is kinda native with a flag being passed (`--init` or `init: /usr/libexec/docker-init`), but some versions of docker-compose implement it, and some don't (some revisions of docker-compose v3 [dropped support then re-added it](https://stackoverflow.com/a/50357065), that's odd).

That's why I don't like this way and I prefer to keep it within the `su-exec` invocation.

> and the arbitrary 991 UID (I'd prefer it be user customizable, perhaps with an ENV var as you mentioned)

Like I said, it is already configurable, no worries! :)
`UID`/`GID` can indeed be passed as environment variables, `991` is just the default.

>  would it be OK if by default it ran as the root user, but with the easy option of being able to set your own UID/GUID?

Well, even if that was possible, I don't think we should do that. We have to keep in mind many people are running Docker images, docker-compose stacks, without even carefully reading them.

> I ask this because I don't want to override the permissions of folders for users and perhaps prevent access.

By setting a default UID/GID we can avoid a security risk without much inconvenience. If they want to change later their UID/GID, they absolutely can! :)

That being said, I'm ready to re-open this PR if you're interested in this solution. As @SuperSandro2000 was reviewing I was already making some changes (you can see that on my fork if needed), but I realised maybe you'll a "conventional" way more appropriate. To be honest I'm not following every official guideline, all I can do is suggest what works for me the best.

I'll take a look at that later, but I'm sure we'll find something we can all agree on. :)

**Comment by Tzahi12345** [comment]
>That's why I don't like this way and I prefer to keep the init within the su-exec invocation.

That inconsistency with Docker compose versions is concerning, perhaps using tini w/ su-exec is the best way.

>Like I said, it is already configurable, no worries! :)

Must have missed that, awesome!

>That being said, I'm ready to re-open this PR if you're interested in this solution. As @SuperSandro2000 was reviewing I was already making some changes (you can see that on my fork if needed), but I realised maybe you'll find a "conventional" way more appropriate. To be honest I'm not following every official guideline, all I can do is suggest what works for me the best.

Here's my take on this: I'm happy taking your solution as a stopgap (at least). If @SuperSandro2000 can re-implement with a more conventional approach, that would be great too. I don't want to continue with this root-only practice because it might drive some users away, and that's the last thing I want to do. If you can branch off your master and remove the Alpine update for simplicity's sake, I would appreciate that.

Let me know if you're willing to do that, I'll take a closer look this evening once I get some free time. Thanks!

**Comment by SuperSandro2000** [comment]
> some revisions of docker-compose v3 dropped support then re-added it, that's odd

revision 3 is for swarm and they removed a lot of features compared to 2. If you are no using swarm you should use version 2.

**Comment by Wonderfall** [comment]
I agree with your statement that revision 3 was designed [towards cross-compatibility between Swarm and Compose](https://docs.docker.com/compose/compose-file/compose-versioning/#version-3). But v3 is still the most current version and [the one that is recommended](https://docs.docker.com/compose/compose-file/compose-versioning/) as it did remove some features, but added some in return.

It's not worth it for everyone to migrate from v2 to v3 for sure, but I still suggest a new installation should use the last & recommended version, which is v3.

**Comment by SuperSandro2000** [comment]
V2 still got updates and new features after 3.0 was released so I wouldn't recommend anyone to use v3 for anything other than swarm which I also can't recommend but that is a different topic.

V3 slowly gets all the features from v2 but it is still missing one, in my opinion, very important one: ``extends``. Without this you need some tool to write your yaml file with defaults. In v2 I can just extend a common base to set security defaults to all my container.

And file format v3 already has --init so we don't need to bloat the container with it.

PS: https://wiki.debian.org/DontBreakDebian#Don.27t_suffer_from_Shiny_New_Stuff_Syndrome ;)

**Comment by Wonderfall** [comment]
> V2 still got updates and new features after 3.0 was released so I wouldn't recommend anyone to use v3 for anything other than swarm which I also can't recommend but that is a different topic.

Like I said, I understand your opinion and I respect it. But I don't see why it's reasonable in this case to go against official recommendations just because you think Swarm is not a solution you like, not to mention v3 isn't just reserved to Swarm (but you know that).

> PS: https://wiki.debian.org/DontBreakDebian#Don.27t_suffer_from_Shiny_New_Stuff_Syndrome ;)

I don't think it's relevant in this discussion. And I don't agree with some points of it, but that's a whole another subject (but a very interesting one!).

> And file format v3 already has --init 

That was not my argument, it was indeed added back to v3. But some revisions don't have it. Really that's not a big deal, thats was just one of the many reasons I'm not a fan of `--init`. Since this is only my opinion, I don't mind if YoutubeDL choose another way of implementing `tini`.

> so we don't need to bloat the container with it.

That's a bit unfair, `tini` has no dependencies and has a installed size of 32kb. Strictly speaking, you could see it at a redundancy, sure. :(

----

Look, I said that from the beginning : my PR is an attempt to fix an important issue with security implications. It works well for me and I hope it can serve people in the same need in the meantime.

I can gladly talk about details, but that's not the point, and that's not the time. As you were reviewing, I was already making some changes, like using an `ENTRYPOINT`, which I do for most of my images. That's a great suggestion, though that could be done before/after. I can remove `tini` if you insist on it, that's really not a big deal. What matters is I don't want to see people running Docker images with a false sense of security.

Unless you want to make your own changes, I can re-open this PR. That's why I closed it in the first place ; since you're a contributor of this project, I thought you might have a better way. I'm ready to work on it if you think it's okay. :)

**Comment by Tzahi12345** [comment]
@SuperSandro2000 has submitted a PR that addresses this problem, looks like he took some ideas from your PR which is great to see.

@Wonderfall I'm gonna go ahead and merge [his PR](https://github.com/Tzahi12345/YoutubeDL-Material/pull/116) instead, but I can't understate how much I appreciate you working on this PR and (ultimately) getting us to fix this issue that really needed to be fixed. Your contribution here is amazing, and I hope it will ultimately bring more people to the project, or at least drive less people away.

Thank you so much! If you notice anything else we're doing that is stupid/a bad practice, always feel free to open up an issue or better yet a PR :)

**Comment by Wonderfall** [comment]
No problem, I'll keep an eye on your project anyway since me and my users are using it (and we love it).

Glad it's solve!

**Comment by Tzahi12345** [comment]
Cool, great to hear :)
